File: app/src/main/java/dev/ragnarok/fenrir/adapter/horizontal/HorizontalPlaylistAdapter.java
package dev.ragnarok.fenrir.adapter.horizontal;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.BitmapFactory;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.List;

import dev.ragnarok.fenrir.Constants;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.adapter.base.RecyclerBindableAdapter;
import dev.ragnarok.fenrir.model.AudioPlaylist;
import dev.ragnarok.fenrir.picasso.transforms.ImageHelper;
import dev.ragnarok.fenrir.picasso.transforms.PolyTransformation;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.util.AppTextUtils;
import dev.ragnarok.fenrir.util.Utils;
import dev.ragnarok.fenrir.util.ViewUtils;

public class HorizontalPlaylistAdapter extends RecyclerBindableAdapter<AudioPlaylist, HorizontalPlaylistAdapter.Holder> {

    private Listener listener;

    public HorizontalPlaylistAdapter(List<AudioPlaylist> data) {
        super(data);
    }

    @SuppressLint("SetTextI18n")
    @Override
    protected void onBindItemViewHolder(Holder holder, int position, int type) {
        AudioPlaylist playlist = getItem(position);

        Context context = holder.itemView.getContext();

        if (!Utils.isEmpty(playlist.getThumb_image()))
            ViewUtils.displayAvatar(holder.thumb, new PolyTransformation(), playlist.getThumb_image(), Constants.PICASSO_TAG);
        else
            holder.thumb.setImageBitmap(ImageHelper.getEllipseBitmap(BitmapFactory.decodeResource(context.getResources(), Settings.get().ui().isDarkModeEnabled(context) ? R.drawable.generic_audio_nowplaying_dark : R.drawable.generic_audio_nowplaying_light), 0.1f));
        holder.count.setText(playlist.getCount() + " " + context.getString(R.string.audios_pattern_count));
        holder.name.setText(playlist.getTitle());
        if (Utils.isEmpty(playlist.getDescription()))
            holder.description.setVisibility(View.GONE);
        else {
            holder.description.setVisibility(View.VISIBLE);
            holder.description.setText(playlist.getDescription());
            holder.description.setOnClickListener(view -> {
                if (holder.description.getMaxLines() == 1)
                    holder.description.setMaxLines(6);
                else
                    holder.description.setMaxLines(1);
            });
        }
        if (Utils.isEmpty(playlist.getArtist_name()))
            holder.artist.setVisibility(View.GONE);
        else {
            holder.artist.setVisibility(View.VISIBLE);
            holder.artist.setText(playlist.getArtist_name());
        }
        if (playlist.getYear() == 0)
            holder.year.setVisibility(View.GONE);
        else {
            holder.year.setVisibility(View.VISIBLE);
            holder.year.setText(String.valueOf(playlist.getYear()));
        }
        if (Utils.isEmpty(playlist.getGenre()))
            holder.genre.setVisibility(View.GONE);
        else {
            holder.genre.setVisibility(View.VISIBLE);
            holder.genre.setText(playlist.getGenre());
        }
        holder.update.setText(AppTextUtils.getDateFromUnixTime(context, playlist.getUpdate_time()));
        holder.add.setOnClickListener(v -> listener.onPlayListClick(playlist, position));
        if (playlist.getOwnerId() == Settings.get().accounts().getCurrent())
            holder.add.setImageResource(R.drawable.ic_outline_delete);
        else
            holder.add.setImageResource(R.drawable.plus);
    }

    @Override
    protected Holder viewHolder(View view, int type) {
        return new Holder(view);
    }

    @Override
    protected int layoutId(int type) {
        return R.layout.item_internal_audio_playlist;
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    public interface Listener {
        void onPlayListClick(AudioPlaylist item, int pos);
    }

    static class Holder extends RecyclerView.ViewHolder {

        final ImageView thumb;
        final TextView name;
        final TextView description;
        final TextView count;
        final TextView year;
        final TextView artist;
        final TextView genre;
        final TextView update;
        final FloatingActionButton add;

        public Holder(View itemView) {
            super(itemView);
            thumb = itemView.findViewById(R.id.item_thumb);
            name = itemView.findViewById(R.id.item_name);
            count = itemView.findViewById(R.id.item_count);
            description = itemView.findViewById(R.id.item_description);
            update = itemView.findViewById(R.id.item_time);
            year = itemView.findViewById(R.id.item_year);
            artist = itemView.findViewById(R.id.item_artist);
            genre = itemView.findViewById(R.id.item_genre);
            add = itemView.findViewById(R.id.add_playlist);
        }
    }
}

File: app/src/main/java/dev/ragnarok/fenrir/adapter/horizontal/HorizontalPlaylistAdapter.kt
package dev.ragnarok.fenrir.adapter.horizontal

import android.annotation.SuppressLint
import android.graphics.BitmapFactory
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.floatingactionbutton.FloatingActionButton
import dev.ragnarok.fenrir.Constants
import dev.ragnarok.fenrir.R
import dev.ragnarok.fenrir.adapter.base.RecyclerBindableAdapter
import dev.ragnarok.fenrir.model.AudioPlaylist
import dev.ragnarok.fenrir.nonNullNoEmpty
import dev.ragnarok.fenrir.picasso.transforms.ImageHelper.getEllipseBitmap
import dev.ragnarok.fenrir.picasso.transforms.PolyTransformation
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.util.AppTextUtils
import dev.ragnarok.fenrir.util.ViewUtils.displayAvatar

class HorizontalPlaylistAdapter(data: MutableList<AudioPlaylist>) :
    RecyclerBindableAdapter<AudioPlaylist, HorizontalPlaylistAdapter.Holder>(data) {
    private var listener: Listener? = null

    @SuppressLint("SetTextI18n")
    override fun onBindItemViewHolder(viewHolder: Holder, position: Int, type: Int) {
        val playlist = getItem(position)
        val context = viewHolder.itemView.context
        if (playlist.thumb_image.nonNullNoEmpty()) displayAvatar(
            viewHolder.thumb,
            PolyTransformation(),
            playlist.thumb_image,
            Constants.PICASSO_TAG
        ) else viewHolder.thumb.setImageBitmap(
            getEllipseBitmap(
                BitmapFactory.decodeResource(
                    context.resources,
                    if (Settings.get().ui()
                            .isDarkModeEnabled(context)
                    ) R.drawable.generic_audio_nowplaying_dark else R.drawable.generic_audio_nowplaying_light
                ), 0.1f
            )
        )
        viewHolder.count.text =
            playlist.count.toString() + " " + context.getString(R.string.audios_pattern_count)
        viewHolder.name.text = playlist.title
        if (playlist.description.isNullOrEmpty()) viewHolder.description.visibility =
            View.GONE else {
            viewHolder.description.visibility = View.VISIBLE
            viewHolder.description.text = playlist.description
            viewHolder.description.setOnClickListener {
                if (viewHolder.description.maxLines == 1) viewHolder.description.maxLines =
                    6 else viewHolder.description.maxLines = 1
            }
        }
        if (playlist.artist_name.isNullOrEmpty()) viewHolder.artist.visibility = View.GONE else {
            viewHolder.artist.visibility = View.VISIBLE
            viewHolder.artist.text = playlist.artist_name
        }
        if (playlist.year == 0) viewHolder.year.visibility = View.GONE else {
            viewHolder.year.visibility = View.VISIBLE
            viewHolder.year.text = playlist.year.toString()
        }
        if (playlist.genre.isNullOrEmpty()) viewHolder.genre.visibility = View.GONE else {
            viewHolder.genre.visibility = View.VISIBLE
            viewHolder.genre.text = playlist.genre
        }
        viewHolder.update.text = AppTextUtils.getDateFromUnixTime(context, playlist.update_time)
        viewHolder.add.setOnClickListener { listener?.onPlayListClick(playlist, position) }
        viewHolder.share.setOnClickListener { listener?.onShareClick(playlist, position) }
        if (playlist.id >= 0) {
            viewHolder.add.visibility = View.VISIBLE
            viewHolder.share.visibility = View.VISIBLE
            if (playlist.ownerId == Settings.get()
                    .accounts().current
            ) viewHolder.add.setImageResource(R.drawable.ic_outline_delete) else viewHolder.add.setImageResource(
                R.drawable.plus
            )
        } else {
            viewHolder.add.visibility = View.GONE
            viewHolder.share.visibility = View.GONE
        }
    }

    override fun viewHolder(view: View, type: Int): Holder {
        return Holder(view)
    }

    override fun layoutId(type: Int): Int {
        return R.layout.item_internal_audio_playlist
    }

    fun setListener(listener: Listener?) {
        this.listener = listener
    }

    interface Listener {
        fun onPlayListClick(item: AudioPlaylist, pos: Int)
        fun onShareClick(item: AudioPlaylist, pos: Int)
    }

    class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val thumb: ImageView = itemView.findViewById(R.id.item_thumb)
        val name: TextView = itemView.findViewById(R.id.item_name)
        val description: TextView = itemView.findViewById(R.id.item_description)
        val count: TextView = itemView.findViewById(R.id.item_count)
        val year: TextView = itemView.findViewById(R.id.item_year)
        val artist: TextView = itemView.findViewById(R.id.item_artist)
        val genre: TextView = itemView.findViewById(R.id.item_genre)
        val update: TextView = itemView.findViewById(R.id.item_time)
        val add: FloatingActionButton = itemView.findViewById(R.id.add_playlist)
        val share: FloatingActionButton = itemView.findViewById(R.id.share_playlist)
    }
}